### PR TITLE
Specify type support for indices and ranges

### LINKED ISSUE #14449: Indices and ranges: supported sequence types
Not only the mentioned array and Span types support indices and ranges.

https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/index-range-changes.md (which is about to be merged into https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md) introduces the notion of countable types. That way, the support of indices and ranges might be pattern-based.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: a92f2fd3-888f-340e-72dd-8a3338313783
* Version Independent ID: 67981ff3-5016-3147-b3c2-5bc16efe1416
* Content: [Explore ranges of data using indices and ranges](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/ranges-indexes)
* Content Source: [docs/csharp/tutorials/ranges-indexes.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tutorials/ranges-indexes.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by pkulikov**
Fixes #14449
Relies on dotnet/csharplang#2797

### CODE REVIEW

File: docs/csharp/tutorials/ranges-indexes.md [hunk 1]
```diff
@@ -69,7 +69,15 @@ The following sample shows many of the reasons for those choices. Modify `x`, `y
 
 [!code-csharp[SemanticsExamples](~/samples/csharp/tutorials/RangesIndexes/IndicesAndRanges.cs#IndicesAndRanges_Semantics)]
 
-## Scenarios for Indices and Ranges
+## Type support for indices and ranges
+
+If a type provides an [indexer](../programming-guide/indexers/index.md) with an <xref:System.Index> or <xref:System.Range> parameter, it explicitly supports indices or ranges respectively.
+
+A type is **countable** if it has a property named `Length` or `Count` with an accessible getter and a return type of `int`. A countable type that doesn't explicitly support indices or ranges may provide an implicit support for them. For more information, see the [Implicit Index support](~/_csharplang/proposals/csharp-8.0/ranges.md#implicit-index-support) and [Implicit Range support](~/_csharplang/proposals/csharp-8.0/ranges.md#implicit-range-support) sections of the [feature proposal note](~/_csharplang/proposals/csharp-8.0/ranges.md).
```
Comment (pkulikov) [inline docs/csharp/tutorials/ranges-indexes.md]: The links in this paragraph assume that  dotnet/csharplang#2797 is merged. Is that fine?

**Comment by pkulikov** [comment]
@BillWagner as the feature proposal note is clear, I try to introduce as little duplication as possible and refer to the note sections. Is that fine or you think the docs should be more vocal?

File: docs/csharp/tutorials/ranges-indexes.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/tutorials/ranges-indexes.md]: Yes, that's fine.  I'll wait to merge this until the related PR is merged. I'm hoping that happens soon. There are some delays as teams are focused on .NET Conf.

**Review by BillWagner — APPROVED** [review]
This is great work @pkulikov 

I've approved the PR, and I'll :shipit: once dotnet/csharplang#2797 is merged.

**Comment by pkulikov** [comment]
@BillWagner the related PR has been merged

**Pull request merged**